pkg/runtime/controller.go

```diff
@@ -0,0 +1,420 @@
+	dispatcher ResolverChangeListener
+
+	// handlerPool is the goroutine pool used by handler.
+	handlerPool *pool.GoroutinePool
+
+	// resolverID to be assigned to the next resolver.
+	// It is incremented every call.
+	resolverID int
+
+	// Fields below are used for testing.
+
+	// createHandlerFactory for testing.
+	createHandlerFactory factoryCreatorFunc
+}
+
+// RulesKind defines the config kind name of mixer rules.
+const RulesKind = "mixer-rules"
+
+// ResolverChangeListener is notified when a new resolver is created due to config change.
+type ResolverChangeListener interface {
+	ChangeResolver(rt Resolver)
+}
+
+// factoryCreatorFunc creates a handler factory. It is used for testing.
+type factoryCreatorFunc func(templateInfo map[string]template.Info, expr expr.TypeChecker,
+	df expr.AttributeDescriptorFinder, builderInfo map[string]*adapter.BuilderInfo) HandlerFactory
+
+// applyEventsFn is used for testing
+type applyEventsFn func(events []*store.Event)
+
+// maxEvents is the likely maximum number of events
+// we can expect in a second. It is used to avoid slice reallocation.
+const maxEvents = 50
+
+// publishSnapShot converts the currently available configState into a resolver.
+// The config may be in an inconsistent state, however it *must* be converted into a consistent resolver.
+// The previous handler table enables handler cleanup and reuse.
+// This code is single threaded, it only runs on a config change control loop.
+func (c *Controller) publishSnapShot() {
+	// current consistent view of handler configuration
+	// keyed by Name.Kind.NameSpace
+	handlerConfig := c.validHandlerConfigs()
+
+	// current consistent view of the Instance configuration
+	// keyed by Name.Kind.NameSpace
+	instanceConfig := c.validInstanceConfigs()
+
+	// new handler factory is created for every config change.
+	hb := c.createHandlerFactory(c.templateInfo, c.eval, c.attrDescFinder, c.adapterInfo)
+
+	// new handler table is created for every config change. It uses handler factory
+	// to create new handlers.
+	ht := newHandlerTable(instanceConfig, handlerConfig,
+		func(handler *cpb.Handler, instances []*cpb.Instance) (adapter.Handler, error) {
+			return hb.Build(handler, instances, newEnv(handler.Name, c.handlerPool))
+		},
+	)
+
+	// current consistent view of the rules keyed by Namespace and then Name.
+	// ht (handlerTable) keeps track of handler-instance association.
+	ruleConfig := c.processRules(handlerConfig, instanceConfig, ht)
+
+	// Initialize handlers that are used in the configuration.
+	// Some handlers may not initialize due to errors.
+	ht.Initialize(c.table)
+
+	// Combine rules with the handler table.
+	// Actions referring to handlers in error, will be purged.
+	combineRulesHandlers(ruleConfig, ht.table)
+
+	// create rules that are in the format that resolver needs.
+	rules, nrules := convertToRuntimeRules(ruleConfig)
+
+	// Create new resolver and cleanup the old resolver.
+	c.resolverID++
```

Comment:
Sorry, I meant, I am not sure how will resolveId be used... It is mainly for testing in order to inspect the value ?